File: app/src/main/java/com/iraguzov/pomodo/data/History.kt
package com.iraguzov.pomodo.data

import android.content.Context
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import org.json.JSONArray
import org.json.JSONObject
import java.io.File
import java.time.DayOfWeek
import java.time.Instant
import java.time.LocalDate
import java.time.ZoneId
import java.time.temporal.TemporalAdjusters

/** Один прогон таймера: сколько собирались и сколько получилось на самом деле. */
data class Session(
    val mode: Mode,
    val startedAt: Long,
    val plannedMs: Long,
    val elapsedMs: Long,
) {
    val workedMs: Long get() = minOf(elapsedMs, plannedMs)
    val overtimeMs: Long get() = (elapsedMs - plannedMs).coerceAtLeast(0L)
}

/**
 * История лежит в приватном файле приложения — ничего никуда не уходит.
 * Формат простой JSON, чтобы не тянуть базу ради пары полей.
 */
class HistoryStore(context: Context) {

    private val file = File(context.applicationContext.filesDir, "history.json")

    suspend fun load(): List<Session> = withContext(Dispatchers.IO) { read() }

    suspend fun append(session: Session): List<Session> = withContext(Dispatchers.IO) {
        appendBlocking(session)
    }

    /** Нужен, когда приложение уже закрывается и корутины запускать поздно. */
    fun appendBlocking(session: Session): List<Session> {
        val sessions = (read() + session).takeLast(MAX_SESSIONS)
        write(sessions)
        return sessions
    }

    suspend fun clear(): List<Session> = withContext(Dispatchers.IO) {
        write(emptyList())
        emptyList()
    }

    private fun read(): List<Session> {
        if (!file.exists()) return emptyList()
        return runCatching {
            val array = JSONArray(file.readText())
            (0 until array.length()).mapNotNull { index ->
                val item = array.getJSONObject(index)
                val mode = runCatching { Mode.valueOf(item.getString("mode")) }.getOrNull()
                    ?: return@mapNotNull null
                Session(
                    mode = mode,
                    startedAt = item.getLong("startedAt"),
                    plannedMs = item.getLong("plannedMs"),
                    elapsedMs = item.getLong("elapsedMs"),
                )
            }
        }.getOrDefault(emptyList())
    }

    private fun write(sessions: List<Session>) {
        val array = JSONArray()
        sessions.forEach { session ->
            array.put(
                JSONObject().apply {
                    put("mode", session.mode.name)
                    put("startedAt", session.startedAt)
                    put("plannedMs", session.plannedMs)
                    put("elapsedMs", session.elapsedMs)
                },
            )
        }
        runCatching { file.writeText(array.toString()) }
    }

    private companion object {
        const val MAX_SESSIONS = 2000
    }
}

enum class StatsPeriod(val label: String) {
    DAY("День"),
    WEEK("Неделя"),
    MONTH("Месяц"),
}

data class Totals(val workedMs: Long, val overtimeMs: Long, val restMs: Long) {
    val isEmpty: Boolean get() = workedMs == 0L && overtimeMs == 0L && restMs == 0L
    val maxMs: Long get() = maxOf(workedMs + overtimeMs, restMs)
}

/** Столбик графика: короткая подпись под осью, полная — для подсказки, плюс те же три числа. */
data class Bucket(val label: String, val title: String, val emphasised: Boolean, val totals: Totals)

private fun totalsOf(sessions: List<Session>): Totals = Totals(
    workedMs = sessions.filter { it.mode == Mode.WORK }.sumOf { it.workedMs },
    overtimeMs = sessions.filter { it.mode == Mode.WORK }.sumOf { it.overtimeMs },
    restMs = sessions.filter { it.mode == Mode.BREAK }.sumOf { it.elapsedMs },
)

/** Периоды календарные: день, неделя с понедельника или месяц, в который попадает дата. */
fun startOf(period: StatsPeriod, date: LocalDate): LocalDate = when (period) {
    StatsPeriod.DAY -> date
    StatsPeriod.WEEK -> date.with(TemporalAdjusters.previousOrSame(DayOfWeek.MONDAY))
    StatsPeriod.MONTH -> date.withDayOfMonth(1)
}

/** Первый день следующего периода — граница не включается. */
private fun endOf(period: StatsPeriod, date: LocalDate): LocalDate = shift(period, startOf(period, date), 1)

fun shift(period: StatsPeriod, date: LocalDate, steps: Long): LocalDate = when (period) {
    StatsPeriod.DAY -> date.plusDays(steps)
    StatsPeriod.WEEK -> date.plusWeeks(steps)
    StatsPeriod.MONTH -> date.plusMonths(steps)
}

/** Вперёд листать можно, пока не дошли до периода с сегодняшним днём. */
fun canShiftForward(period: StatsPeriod, date: LocalDate, zone: ZoneId = ZoneId.systemDefault()): Boolean =
    !endOf(period, date).isAfter(LocalDate.now(zone))

fun sessionsIn(
    sessions: List<Session>,
    period: StatsPeriod,
    date: LocalDate,
    zone: ZoneId = ZoneId.systemDefault(),
): List<Session> {
    val from = startOf(period, date)
    val until = endOf(period, date)
    return sessions.filter { session ->
        val day = dateOf(session, zone)
        !day.isBefore(from) && day.isBefore(until)
    }
}

fun totalsFor(
    sessions: List<Session>,
    period: StatsPeriod,
    date: LocalDate,
    zone: ZoneId = ZoneId.systemDefault(),
): Totals = totalsOf(sessionsIn(sessions, period, date, zone))

/**
 * Прогон попадает в корзину по времени своего начала — сессии короткие,
 * и делить их между часами ради точности не стоит.
 */
fun bucketsFor(
    sessions: List<Session>,
    period: StatsPeriod,
    date: LocalDate,
    zone: ZoneId = ZoneId.systemDefault(),
): List<Bucket> {
    val relevant = sessionsIn(sessions, period, date, zone)
    return when (period) {
        StatsPeriod.DAY -> (0..23).map { hour ->
            Bucket(
                label = "%02d".format(hour),
                title = "%02d:00–%02d:00".format(hour, (hour + 1) % 24),
                emphasised = hour % 6 == 0,
                totals = totalsOf(relevant.filter { hourOf(it, zone) == hour }),
            )
        }

        else -> {
            val until = endOf(period, date)
            generateSequence(startOf(period, date)) { it.plusDays(1) }
                .takeWhile { it.isBefore(until) }
                .map { day ->
                    Bucket(
                        label = if (period == StatsPeriod.WEEK) weekdayOf(day) else day.dayOfMonth.toString(),
                        title = dayTitle(day),
                        emphasised = period == StatsPeriod.WEEK || day.dayOfMonth == 1 || day.dayOfMonth % 5 == 0,
                        totals = totalsOf(relevant.filter { dateOf(it, zone) == day }),
                    )
                }
                .toList()
        }
    }
}

/** «Сегодня», «8–14 сентября», «Август»; год добавляется, только если он не текущий. */
fun periodTitle(period: StatsPeriod, date: LocalDate, zone: ZoneId = ZoneId.systemDefault()): String {
    val today = LocalDate.now(zone)
    fun year(day: LocalDate) = if (day.year == today.year) "" else " ${day.year}"
    return when (period) {
        StatsPeriod.DAY -> when (date) {
            today -> "Сегодня"
            today.minusDays(1) -> "Вчера"
            else -> dayTitle(date) + year(date)
        }

        StatsPeriod.WEEK -> {
            val first = startOf(period, date)
            val last = first.plusDays(6)
            if (first.month == last.month) {
                "${first.dayOfMonth}–${last.dayOfMonth} ${monthOf(last)}${year(last)}"
            } else {
                "${first.dayOfMonth} ${monthOf(first)} – ${last.dayOfMonth} ${monthOf(last)}${year(last)}"
            }
        }

        StatsPeriod.MONTH -> MonthsNominative[date.monthValue - 1] + year(date)
    }
}

// Интерфейс русский, поэтому даты подписываем по-русски независимо от языка системы.
private val WeekdaysShort = listOf("Пн", "Вт", "Ср", "Чт", "Пт", "Сб", "Вс")
private val MonthsGenitive = listOf(
    "января", "февраля", "марта", "апреля", "мая", "июня",
    "июля", "августа", "сентября", "октября", "ноября", "декабря",
)
private val MonthsNominative = listOf(
    "Январь", "Февраль", "Март", "Апрель", "Май", "Июнь",
    "Июль", "Август", "Сентябрь", "Октябрь", "Ноябрь", "Декабрь",
)

private fun weekdayOf(day: LocalDate): String = WeekdaysShort[day.dayOfWeek.value - 1]

private fun monthOf(day: LocalDate): String = MonthsGenitive[day.monthValue - 1]

/** «Пн, 8 сентября». */
private fun dayTitle(day: LocalDate): String = "${weekdayOf(day)}, ${day.dayOfMonth} ${monthOf(day)}"

private fun dateOf(session: Session, zone: ZoneId): LocalDate =
    Instant.ofEpochMilli(session.startedAt).atZone(zone).toLocalDate()

private fun hourOf(session: Session, zone: ZoneId): Int =
    Instant.ofEpochMilli(session.startedAt).atZone(zone).hour

/** Шаг сетки графика: круглое число минут или часов, чтобы делений было не больше четырёх. */
fun axisStepMs(peakMs: Long): Long {
    val stepMinutes = longArrayOf(1, 2, 5, 10, 15, 30, 60, 120, 180, 240, 360, 720)
    stepMinutes.forEach { minutes ->
        val step = minutes * 60_000
        if (peakMs <= step * 4) return step
    }
    val hour = 3_600_000L
    return (peakMs + hour * 4 - 1) / (hour * 4) * hour
}

/** Подпись деления оси: «0», «15 мин», «1 ч», «1,5 ч». */
fun formatAxis(ms: Long): String {
    val minutes = ms / 60_000
    return when {
        minutes == 0L -> "0"
        minutes < 60 -> "$minutes мин"
        minutes % 60 == 0L -> "${minutes / 60} ч"
        minutes % 30 == 0L -> "${minutes / 60},5 ч"
        else -> "${minutes / 60} ч ${"%02d".format(minutes % 60)}"
    }
}

/** «1 ч 05 мин», «12 мин», «—» для пустого значения. */
fun formatSpan(ms: Long): String {
    if (ms <= 0L) return "—"
    val totalMinutes = ms / 60_000
    val hours = totalMinutes / 60
    val minutes = totalMinutes % 60
    return when {
        hours > 0 -> "$hours ч ${"%02d".format(minutes)} мин"
        minutes > 0 -> "$minutes мин"
        else -> "меньше минуты"
    }
}

File: ios/Pomodo/History.swift
import Foundation

/// Один прогон таймера: сколько собирались и сколько получилось на самом деле.
struct Session: Codable, Identifiable {
    let mode: Mode
    let startedAt: Date
    let plannedMs: Int
    let elapsedMs: Int

    var id: Date { startedAt }
    var workedMs: Int { min(elapsedMs, plannedMs) }
    var overtimeMs: Int { max(elapsedMs - plannedMs, 0) }
}

/// История лежит в приватном файле приложения — ничего никуда не уходит.
@Observable
final class HistoryStore {

    private(set) var sessions: [Session] = []

    @ObservationIgnored private let url: URL
    @ObservationIgnored private let maxSessions = 2000

    init() {
        let directory = FileManager.default.urls(for: .applicationSupportDirectory, in: .userDomainMask)[0]
        try? FileManager.default.createDirectory(at: directory, withIntermediateDirectories: true)
        url = directory.appendingPathComponent("history.json")
        load()
    }

    func append(_ session: Session) {
        sessions = Array((sessions + [session]).suffix(maxSessions))
        save()
    }

    func clear() {
        sessions = []
        save()
    }

    private func load() {
        guard let data = try? Data(contentsOf: url) else { return }
        sessions = (try? JSONDecoder().decode([Session].self, from: data)) ?? []
    }

    private func save() {
        guard let data = try? JSONEncoder().encode(sessions) else { return }
        try? data.write(to: url, options: .atomic)
    }
}

enum StatsPeriod: String, CaseIterable, Identifiable {
    case day, week, month

    var id: String { rawValue }

    var label: String {
        switch self {
        case .day: "День"
        case .week: "Неделя"
        case .month: "Месяц"
        }
    }

    fileprivate var component: Calendar.Component {
        switch self {
        case .day: .day
        case .week: .weekOfYear
        case .month: .month
        }
    }
}

struct Totals {
    let workedMs: Int
    let overtimeMs: Int
    let restMs: Int

    var isEmpty: Bool { workedMs == 0 && overtimeMs == 0 && restMs == 0 }
    var maxMs: Int { max(workedMs + overtimeMs, restMs) }
}

/// Столбик графика: короткая подпись под осью, полная — для подсказки, плюс те же три числа.
struct Bucket: Identifiable {
    let id: Int
    let label: String
    let title: String
    let emphasised: Bool
    let totals: Totals
}

private func totalsOf(_ sessions: [Session]) -> Totals {
    Totals(
        workedMs: sessions.filter { $0.mode == .work }.reduce(0) { $0 + $1.workedMs },
        overtimeMs: sessions.filter { $0.mode == .work }.reduce(0) { $0 + $1.overtimeMs },
        restMs: sessions.filter { $0.mode == .rest }.reduce(0) { $0 + $1.elapsedMs }
    )
}

/// Интерфейс русский, поэтому неделя начинается с понедельника независимо от настроек системы.
var statsCalendar: Calendar {
    var calendar = Calendar(identifier: .gregorian)
    calendar.locale = Locale(identifier: "ru_RU")
    calendar.firstWeekday = 2
    return calendar
}

/// Периоды календарные: день, неделя с понедельника или месяц, в который попадает дата.
func range(of period: StatsPeriod, containing date: Date) -> DateInterval {
    statsCalendar.dateInterval(of: period.component, for: date) ?? DateInterval(start: date, duration: 0)
}

func shifted(_ date: Date, by steps: Int, period: StatsPeriod) -> Date {
    statsCalendar.date(byAdding: period.component, value: steps, to: date) ?? date
}

/// Вперёд листать можно, пока не дошли до периода с сегодняшним днём.
func canShiftForward(_ date: Date, period: StatsPeriod) -> Bool {
    range(of: period, containing: date).end <= Date()
}

func sessionsIn(_ sessions: [Session], period: StatsPeriod, containing date: Date) -> [Session] {
    let interval = range(of: period, containing: date)
    return sessions.filter { $0.startedAt >= interval.start && $0.startedAt < interval.end }
}

func totalsFor(_ sessions: [Session], period: StatsPeriod, containing date: Date) -> Totals {
    totalsOf(sessionsIn(sessions, period: period, containing: date))
}

/// Прогон попадает в корзину по времени своего начала — сессии короткие,
/// и делить их между часами ради точности не стоит.
func bucketsFor(_ sessions: [Session], period: StatsPeriod, containing date: Date) -> [Bucket] {
    let calendar = statsCalendar
    let relevant = sessionsIn(sessions, period: period, containing: date)

    if period == .day {
        return (0...23).map { hour in
            let inHour = relevant.filter { calendar.component(.hour, from: $0.startedAt) == hour }
            return Bucket(
                id: hour,
                label: String(format: "%02d", hour),
                title: String(format: "%02d:00–%02d:00", hour, (hour + 1) % 24),
                emphasised: hour % 6 == 0,
                totals: totalsOf(inHour)
            )
        }
    }

    let interval = range(of: period, containing: date)
    var days: [Date] = []
    var day = interval.start
    while day < interval.end {
        days.append(day)
        day = calendar.date(byAdding: .day, value: 1, to: day) ?? interval.end
    }
    return days.enumerated().map { index, day in
        let inDay = relevant.filter { calendar.isDate($0.startedAt, inSameDayAs: day) }
        let dayOfMonth = calendar.component(.day, from: day)
        return Bucket(
            id: index,
            label: period == .week ? weekdayOf(day) : "\(dayOfMonth)",
            title: dayTitle(day),
            emphasised: period == .week || dayOfMonth == 1 || dayOfMonth % 5 == 0,
            totals: totalsOf(inDay)
        )
    }
}

/// «Сегодня», «8–14 сентября», «Август»; год добавляется, только если он не текущий.
func periodTitle(_ period: StatsPeriod, containing date: Date) -> String {
    let calendar = statsCalendar
    let currentYear = calendar.component(.year, from: Date())
    func year(_ day: Date) -> String {
        let value = calendar.component(.year, from: day)
        return value == currentYear ? "" : " \(value)"
    }

    switch period {
    case .day:
        if calendar.isDateInToday(date) { return "Сегодня" }
        if calendar.isDateInYesterday(date) { return "Вчера" }
        return dayTitle(date) + year(date)
    case .week:
        let interval = range(of: .week, containing: date)
        let first = interval.start
        let last = calendar.date(byAdding: .day, value: -1, to: interval.end) ?? first
        let firstDay = calendar.component(.day, from: first)
        let lastDay = calendar.component(.day, from: last)
        if calendar.isDate(first, equalTo: last, toGranularity: .month) {
            return "\(firstDay)–\(lastDay) \(monthOf(last))" + year(last)
        }
        return "\(firstDay) \(monthOf(first)) – \(lastDay) \(monthOf(last))" + year(last)
    case .month:
        return monthsNominative[calendar.component(.month, from: date) - 1] + year(date)
    }
}

// Интерфейс русский, поэтому даты подписываем по-русски независимо от языка системы.
private let weekdaysShort = ["Пн", "Вт", "Ср", "Чт", "Пт", "Сб", "Вс"]
private let monthsGenitive = [
    "января", "февраля", "марта", "апреля", "мая", "июня",
    "июля", "августа", "сентября", "октября", "ноября", "декабря",
]
private let monthsNominative = [
    "Январь", "Февраль", "Март", "Апрель", "Май", "Июнь",
    "Июль", "Август", "Сентябрь", "Октябрь", "Ноябрь", "Декабрь",
]

private func weekdayOf(_ day: Date) -> String {
    // В Calendar воскресенье — 1, понедельник — 2.
    weekdaysShort[(statsCalendar.component(.weekday, from: day) + 5) % 7]
}

private func monthOf(_ day: Date) -> String {
    monthsGenitive[statsCalendar.component(.month, from: day) - 1]
}

/// «Пн, 8 сентября».
private func dayTitle(_ day: Date) -> String {
    "\(weekdayOf(day)), \(statsCalendar.component(.day, from: day)) \(monthOf(day))"
}

/// Шаг сетки графика: круглое число минут или часов, чтобы делений было не больше четырёх.
func axisStepMs(_ peakMs: Int) -> Int {
    for minutes in [1, 2, 5, 10, 15, 30, 60, 120, 180, 240, 360, 720] {
        let step = minutes * 60_000
        if peakMs <= step * 4 { return step }
    }
    let hour = 3_600_000
    return (peakMs + hour * 4 - 1) / (hour * 4) * hour
}

/// Подпись деления оси: «0», «15 мин», «1 ч», «1,5 ч».
func formatAxis(_ ms: Int) -> String {
    let minutes = ms / 60_000
    if minutes == 0 { return "0" }
    if minutes < 60 { return "\(minutes) мин" }
    if minutes % 60 == 0 { return "\(minutes / 60) ч" }
    if minutes % 30 == 0 { return "\(minutes / 60),5 ч" }
    return "\(minutes / 60) ч \(String(format: "%02d", minutes % 60))"
}

/// «1 ч 05 мин», «12 мин», «—» для пустого значения.
func formatSpan(_ ms: Int) -> String {
    guard ms > 0 else { return "—" }
    let totalMinutes = ms / 60_000
    let hours = totalMinutes / 60
    let minutes = totalMinutes % 60
    if hours > 0 { return "\(hours) ч \(String(format: "%02d", minutes)) мин" }
    if minutes > 0 { return "\(minutes) мин" }
    return "меньше минуты"
}
